File: src/xyplot.py
from dataclasses import dataclass
from typing import Iterator

from box import Box, Margins
from meta import value_with_default, with_config
from plot_ticks import PlotTicks, PlotTicksConfig
from point_transformer import PointTransformer, MoveAndScale
from shape import Circle, Line, Rect, Shape, ShapeParams, ShapeStream, Text, shape_generator
from shape_transformer import ShapeTransformer


@dataclass(frozen=True, kw_only=True)
class XYPlot:
  domain: Box = Box()
  output_range: Box = Box()
  margins: Margins | None = None

  x_axis_values: PlotTicksConfig = PlotTicksConfig()
  y_axis_values: PlotTicksConfig = PlotTicksConfig()

  x_label: str | None = None
  y_label: str | None = None

  labels: frozenset[str] = frozenset()

  identity_line: bool | None = None

  @property
  def transformer(self) -> ShapeTransformer:
    assert self.domain is not None
    assert self.output_range is not None
    output = self.output_range.with_y_reversed()
    if self.margins:
      output = output.with_margins(self.margins)
    return ShapeTransformer(MoveAndScale(self.domain, output))

  def with_defaults(self, defaults: "XYPlot") -> "XYPlot":
    return XYPlot(
        domain=self.domain.with_defaults(defaults.domain),
        output_range=self.output_range.with_defaults(defaults.output_range),
        margins=self.margins or defaults.margins,
        x_axis_values=self.x_axis_values.with_defaults(defaults.x_axis_values),
        y_axis_values=self.y_axis_values.with_defaults(defaults.y_axis_values),
        x_label=self.x_label or defaults.x_label,
        y_label=self.y_label or defaults.y_label,
        labels=self.labels or defaults.labels,
        identity_line=value_with_default(self.identity_line,
                                         defaults.identity_line))

  def produce(self) -> ShapeStream:
    return self.transformer(self._draw()) + self._legend()

  def _legend(self) -> Iterator[Shape]:
    assert self.output_range is not None
    for i, key in enumerate(sorted(self.labels)):
      lx = self.output_range.width() - 60
      ly = 20 + (i * 20)
      yield Rect(lx, ly, 10, 10, ShapeParams(css_class=key))
      yield Text(key, lx + 15, ly + 9)

    if self.x_label:
      yield Text(self.x_label,
                 self.output_range.width() / 2, self.output_range.height(),
                 ShapeParams(css_class="label-x"))
    if self.y_label:
      yield Text(
          self.y_label, 15,
          self.output_range.height() / 2,
          ShapeParams(
              css_class="label-y",
              transform=f"rotate(-90 15,{self.output_range.height()/2})"))

  @shape_generator
  def _draw(self) -> Iterator[Shape]:
    assert self.domain is not None
    assert self.output_range is not None
    assert self.domain.x1 is not None
    assert self.domain.x2 is not None
    assert self.domain.y1 is not None
    assert self.domain.y2 is not None
    x_values = self.x_axis_values.build(self.domain.x1, self.domain.x2)
    for x in x_values.values:
      yield Line.vertical(x, self.domain.y1, self.domain.y2,
                          ShapeParams(css_class="tic"))
      span = (self.domain.height() / 50) * (
          self.output_range.width() / self.output_range.height())
      yield Line.vertical(x, -span, 0)
      yield Text(f"{x:{x_values.value_format}}", x, 2 * -span,
                 ShapeParams(css_class="tic-value-x"))

    y_values = self.y_axis_values.build(self.domain.y1, self.domain.y2)
    for y in y_values.values:
      yield Line.horizontal(self.domain.x1, self.domain.x2, y,
                            ShapeParams(css_class="tic"))
      span = self.domain.width() / 50
      yield Line.horizontal(self.domain.x1 - span, self.domain.x1, y)
      yield Text(f"{y:{y_values.value_format}}", self.domain.x1 - 2 * span, y,
                 ShapeParams(css_class="tic-value-y"))

    yield Line.vertical(self.domain.x1, self.domain.y1, self.domain.y2)
    yield Line.horizontal(self.domain.x1, self.domain.x2, self.domain.y1)

    if self.identity_line:
      clip = min(self.domain.x2, self.domain.y2)
      yield Line(0, 0, clip, clip, ShapeParams(css_class="identity-line"))


with_plot_config = with_config(XYPlot)

File: src/ts/xyplot.ts
import {Box, Margins, simpleBox} from './box.js';
import {getPlotTicks, PlotTicksConfig} from './plot_ticks.js';
import {MoveAndScale, PointTransformer} from './point_transformer.js';
import {Shape, ShapeParams, transformShape} from './shape.js';

export interface XYPlotOptions {
  domain?: Box;
  outputRange?: Box;
  margins?: Margins;
  xAxisValues?: PlotTicksConfig;
  yAxisValues?: PlotTicksConfig;
  xLabel?: string;
  yLabel?: string;
  labels?: Set<string>;
  identityLine?: boolean;
}

export class XYPlot {
  domain: Box;
  outputRange: Box;
  margins?: Margins;
  xAxisValues: PlotTicksConfig;
  yAxisValues: PlotTicksConfig;
  xLabel?: string;
  yLabel?: string;
  labels: Set<string>;
  identityLine?: boolean;

  constructor({
    domain = new Box(),
    outputRange = new Box(),
    margins,
    xAxisValues = {},
    yAxisValues = {},
    xLabel,
    yLabel,
    labels = new Set(),
    identityLine
  }: XYPlotOptions = {}) {
    this.domain = domain;
    this.outputRange = outputRange;
    this.margins = margins;
    this.xAxisValues = xAxisValues;
    this.yAxisValues = yAxisValues;
    this.xLabel = xLabel;
    this.yLabel = yLabel;
    this.labels = labels;
    this.identityLine = identityLine;
  }

  get transformer(): (shapes: Shape[]) => Shape[] {
    let output = this.outputRange.withYReversed();
    if (this.margins) {
      output = output.withMargins(this.margins);
    }
    const pointTransformer = MoveAndScale(this.domain, output);
    return (shapes: Shape[]) =>
               shapes.map(shape => transformShape(shape, pointTransformer));
  }

  withDefaults(defaults: XYPlot): XYPlot {
    return new XYPlot({
      domain: this.domain.withDefaults(defaults.domain),
      outputRange: this.outputRange.withDefaults(defaults.outputRange),
      margins: this.margins ?? defaults.margins,
      xAxisValues: {...defaults.xAxisValues, ...this.xAxisValues},
      yAxisValues: {...defaults.yAxisValues, ...this.yAxisValues},
      xLabel: this.xLabel ?? defaults.xLabel,
      yLabel: this.yLabel ?? defaults.yLabel,
      labels: this.labels.size > 0 ? this.labels : defaults.labels,
      identityLine: this.identityLine ?? defaults.identityLine,
    });
  }

  produce(): Shape[] {
    const drawnShapes = this.draw();
    const transformedShapes = this.transformer(drawnShapes);
    const legendShapes = this.legend();
    return [...transformedShapes, ...legendShapes];
  }

  private legend(): Shape[] {
    const shapes: Shape[] = [];
    const sortedLabels = Array.from(this.labels).sort();

    sortedLabels.forEach((key, i) => {
      const lx = this.outputRange.width() - 60;
      const ly = 20 + i * 20;
      shapes.push({
        type: 'rect',
        x: lx,
        y: ly,
        w: 10,
        h: 10,
        params: {cssClass: `labels-${key}`}
      });
      shapes.push({
        type: 'text',
        text: key,
        x: lx + 15,
        y: ly + 9,
        params: {cssClass: `labels-${key}`}
      });
    });

    if (this.xLabel) {
      shapes.push({
        type: 'text',
        text: this.xLabel,
        x: this.outputRange.width() / 2,
        y: this.outputRange.height(),
        params: {cssClass: 'label-x'}
      });
    }

    if (this.yLabel) {
      const halfHeight = this.outputRange.height() / 2;
      shapes.push({
        type: 'text',
        text: this.yLabel,
        x: 15,
        y: halfHeight,
        params: {cssClass: 'label-y', transform: `rotate(-90 15,${halfHeight})`}
      });
    }

    return shapes;
  }

  private draw(): Shape[] {
    if (this.domain.x1 === null || this.domain.x2 === null ||
        this.domain.y1 === null || this.domain.y2 === null) {
      throw new Error('Domain bounds must be fully defined');
    }

    const shapes: Shape[] = [];

    // X Axis Ticks
    const xValues =
        getPlotTicks(this.xAxisValues, this.domain.x1, this.domain.x2);
    for (const x of xValues.values) {
      shapes.push({
        type: 'line',
        x1: x,
        y1: this.domain.y1,
        x2: x,
        y2: this.domain.y2,
        params: {cssClass: 'grid-line grid-line-x tick tick-x'}
      });
      const span = (this.domain.height() / 50) *
          (this.outputRange.width() / this.outputRange.height());
      shapes.push({
        type: 'line',
        x1: x,
        y1: this.domain.y1 - span,
        x2: x,
        y2: this.domain.y1,
        params: {cssClass: 'tick tick-x'}
      });
      shapes.push({
        type: 'text',
        text: xValues.formatFunction(x),
        x: x,
        y: this.domain.y1 - 2 * span,
        params: {cssClass: 'tick tick-x'}
      });
    }

    // Y Axis Ticks
    const yValues =
        getPlotTicks(this.yAxisValues, this.domain.y1, this.domain.y2);
    for (const y of yValues.values) {
      shapes.push({
        type: 'line',
        x1: this.domain.x1,
        y1: y,
        x2: this.domain.x2,
        y2: y,
        params: {cssClass: 'grid-line grid-line-y tick tick-y'}
      });
      const span = this.domain.width() / 50;
      shapes.push({
        type: 'line',
        x1: this.domain.x1 - span,
        y1: y,
        x2: this.domain.x1,
        y2: y,
        params: {cssClass: 'tick tick-y'}
      });
      shapes.push({
        type: 'text',
        text: yValues.formatFunction(y),
        x: this.domain.x1 - 2 * span,
        y: y,
        params: {cssClass: 'tick tick-y'}
      });
    }

    // Axis Borders
    shapes.push({
      type: 'line',
      x1: this.domain.x1,
      y1: this.domain.y1,
      x2: this.domain.x1,
      y2: this.domain.y2,
      params: {cssClass: 'axis-border axis-border-y'}
    });
    shapes.push({
      type: 'line',
      x1: this.domain.x1,
      y1: this.domain.y1,
      x2: this.domain.x2,
      y2: this.domain.y1,
      params: {cssClass: 'axis-border axis-border-x'}
    });

    // Identity Line
    if (this.identityLine) {
      const clip = Math.min(this.domain.x2, this.domain.y2);
      shapes.push({
        type: 'line',
        x1: 0,
        y1: 0,
        x2: clip,
        y2: clip,
        params: {cssClass: 'identity-line'}
      });
    }

    return shapes;
  }
}
